helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/algorithm/RebalanceAlgorithmFactory.java

```diff
@@ -0,0 +1,52 @@
+package org.apache.helix.controller.rebalancer.waged.algorithm;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.helix.HelixException;
+import org.apache.helix.controller.rebalancer.waged.constraints.ConstraintBasedAlgorithm;
+import org.apache.helix.controller.rebalancer.waged.constraints.HardConstraint;
+import org.apache.helix.controller.rebalancer.waged.constraints.SoftConstraint;
+
+public class RebalanceAlgorithmFactory {
+  private static Map<RebalanceAlgorithm.Type, RebalanceAlgorithm> INSTANCES = new HashMap<>();
+
+  public static synchronized RebalanceAlgorithm getInstance(RebalanceAlgorithm.Type type) {
+    if (INSTANCES.containsKey(type)) {
+      return INSTANCES.get(type);
+    }
+
+    switch (type) {
```

Comment:
Reminder to rid of type if this is no longer necessary?

Reply:
why is this no longer necessary? The purpose of our waged rebalancer is to set up a generic framework so that once we have a better idea of a new algorithm, it can be plugged in easily and I suppose the factory will take care of that.